### PR TITLE
User invite

### PR DESCRIPTION
**Pull request by sgiaccio**

### CODE REVIEW

File: server/modules/user/api/userApi.js [hunk 1]
```diff
@@ -9,6 +9,30 @@ const SystemError = require('../../../../server/utils/systemError')
...
 module.exports.init = app => {
 
+  // ==== CREATE
+
+  app.post('/survey/:surveyId/users/invite', AuthMiddleware.requireUserInvitePermission, async (req, res, next) => {
+    try {
+      const surveyId = Request.getRestParam(req, 'surveyId')
+      const email = Request.getRestParam(req, 'email')
+      const groupId = Request.getRestParam(req, 'groupId')
+
```
Comment (fefrain) [inline server/modules/user/api/userApi.js]: Maybe it's easier to use Request.getParams
E.g. const { surveyId, email, groupId } = Request.getParams(req)

Comment (sgiaccio) [reply server/modules/user/api/userApi.js]: Done

File: common/i18n/resources/en.js [hunk 2]
```diff
@@ -148,6 +148,7 @@ module.exports = {
   usersView: {
     inviteUser: 'Invite',
     accepted: 'Accepted',
+    group: 'Group',
```
Comment (minotogna) [inline common/i18n/resources/en.js]: isn't group already in common ?

File: common/i18n/resources/en.js [hunk 3]
```diff
@@ -343,6 +346,7 @@ module.exports = {
   // ====== Auth
 
   authGroups: {
+    group: 'Group',
```
Comment (minotogna) [inline common/i18n/resources/en.js]: isn't group already in common ?

File: common/user/userValidator.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+const errorKeys = {
+  invalidEmail: 'invalidEmail'
+}
+
+const getProp = (propName, defaultValue) => R.pathOr(defaultValue, propName.split('.'))
+
+const validateEmail = (propName, item) => {
+  const email = getProp(propName)(item)
+  return email && !User.validEmail(email) ? { key: errorKeys.invalidEmail } : null
```
Comment (minotogna) [inline common/user/userValidator.js]: should we move the code User.validEmail here ?

File: server/modules/activityLog/activityLogger.js [hunk 5]
```diff
@@ -38,6 +38,10 @@ const type = {
   nodeCreate: 'nodeCreate',
   nodeValueUpdate: 'nodeValueUpdate',
   nodeDelete: 'nodeDelete',
+
+  // user
```
Comment (minotogna) [inline server/modules/activityLog/activityLogger.js]: why two different activities? shouldn't it be always userInvite? - when user join the application for the first time, then adding a userJoined activity log? what do you think?

File: server/modules/auth/manager/authManager.js [hunk 6]
```diff
@@ -1,6 +1,21 @@
+const db = require('../../../db/db')
+
+const AuthGroupRepository = require('../repository/authGroupRepository')
 const JwtRepository = require('../repository/jwtRepository')
 
+const ActivityLog = require('../../activityLog/activityLogger')
+
+const insertUserGroup = async (user, surveyId, groupId, newUserId, client = db) =>
+  client.tx(async t => {
```
Comment (minotogna) [inline server/modules/auth/manager/authManager.js]: let's not forge the await here

File: server/modules/user/manager/userManager.js [hunk 7]
```diff
@@ -1,35 +1,61 @@
...
+const fetchUserByEmail = _userFetcher(UserRepository.fetchUserByEmail)
+
+const fetchUserByCognitoUsername = _userFetcher(UserRepository.fetchUserByCognitoUsername)
+
 // ==== DELETE
 
 const deleteUserPref = async (user, name) => ({
   ...(await UserRepository.deleteUserPref(user, name)),
-  authGroups: await AuthGroupRepository.fetchUserGroups(user.id)
+  authGroups: await AuthManager.fetchUserGroups(user.id)
```
Comment (minotogna) [inline server/modules/user/manager/userManager.js]: we use the convention that Manager uses Repositories.

File: server/modules/user/manager/userManager.js [hunk 8]
```diff
@@ -1,35 +1,61 @@
...
 // ==== READ
 
-const findUserByCognitoUsername = async email => {
-  const user = await UserRepository.findUserByCognitoUsername(email)
+const _userFetcher = fetchFn => async (...args) => {
+  const user = await fetchFn(...args)
 
   if (user) {
-    const authGroups = await AuthGroupRepository.fetchUserGroups(user.id)
+    const authGroups = await AuthManager.fetchUserGroups(user.id)
```
Comment (minotogna) [inline server/modules/user/manager/userManager.js]: we use the convention that Manager uses Repositories.

File: server/modules/user/manager/userManager.js [hunk 9]
```diff
@@ -1,35 +1,61 @@
...
+
+// ==== CREATE
+
+const insertUser = async (user, surveyId, email, groupId, client = db) =>
+  client.tx(async t => {
+    const newUser = await UserRepository.insertUser(surveyId, email, groupId, t)
+    const newUserId = User.getId(newUser)
+    await ActivityLog.log(user, surveyId, ActivityLog.type.userInvite, { email, id: newUserId }, t)
+
+    await AuthManager.insertUserGroup(user, surveyId, groupId, newUserId, t)
```
Comment (minotogna) [inline server/modules/user/manager/userManager.js]: we use the convention that Manager uses Repositories.

File: server/modules/user/service/userService.js [hunk 10]
```diff
@@ -1,15 +1,51 @@
...
+    if (hasRoleInSurvey) {
+      throw new SystemError('userHasRole')
+    } else {
+      await AuthGroupManager.insertUserGroup(user, surveyId, groupId, newUserId)
+      return dbUser
+    }
+  } else {
+    const newUser = await UserManager.insertUser(user, surveyId, email, groupId)
+    await aws.inviteUser(email)
+
```
Comment (minotogna) [inline server/modules/user/service/userService.js]: why don't we first wait for the successful response from AWS cognito and then insert user into the db?

File: server/system/aws.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
+const aws = require('aws-sdk')
+
+const inviteUser = email =>
+  new Promise((resolve, reject) => {
+    aws.config.update({
```
Comment (minotogna) [inline server/system/aws.js]: is aws.config needed at each inverUser invocation? or can it be done just once?

File: webapp/loggedin/appModules.js [hunk 12]
```diff
@@ -95,4 +95,8 @@ export const userModules = {
     key: 'userList',
     path: `${appModules.users.path}/list`,
   },
+  userInvite: {
+    key: 'inviteUser',
```
Comment (minotogna) [inline webapp/loggedin/appModules.js]: should the key also be userInvite?

File: webapp/loggedin/modules/users/userInvite/userInviteView.js [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  }))
+
+  const { data, error, dispatch } = useAsyncPostRequest(
+    `/api/survey/${surveyId}/users/invite`, {
+      email,
+      groupId: AuthGroups.getId(group)
+    }
+  )
+
+  const inviteUser = async () => {
```
Comment (minotogna) [inline webapp/loggedin/modules/users/userInvite/userInviteView.js]: is ```async``` needed for this function?

File: webapp/loggedin/modules/users/userList/userListView.js [hunk 14]
```diff
@@ -1,19 +1,30 @@
...
+
 import User from '../../../../../common/user/user'
 
-const UsersHeaderLeft = () => {
+import { appModuleUri, userModules } from '../../../appModules'
+
+import * as AppState from '../../../../app/appState'
+import * as SurveyState from '../../../../survey/surveyState'
+
+const UsersHeaderLeft = canInvite => () => {
```
Comment (minotogna) [inline webapp/loggedin/modules/users/userList/userListView.js]: this component doesn't need to be a function that returns the component. you can just pass canInvite prop to the TableView component

File: common/i18n/resources/en.js (hunk 2 above)
Comment (sgiaccio) [reply common/i18n/resources/en.js]: yes, deleted

File: common/i18n/resources/en.js (hunk 3 above)
Comment (sgiaccio) [reply common/i18n/resources/en.js]: yes

File: common/user/userValidator.js (hunk 4 above)
Comment (sgiaccio) [reply common/user/userValidator.js]: Ok, moved

File: server/modules/auth/manager/authManager.js (hunk 6 above)
Comment (sgiaccio) [reply server/modules/auth/manager/authManager.js]: ok

File: webapp/loggedin/appModules.js (hunk 12 above)
Comment (sgiaccio) [reply webapp/loggedin/appModules.js]: yes, changed

File: server/modules/user/service/userService.js (hunk 10 above)
Comment (sgiaccio) [reply server/modules/user/service/userService.js]: Yes, it has already been done on the accept-invitation branch

File: webapp/loggedin/modules/users/userInvite/userInviteView.js (hunk 13 above)
Comment (sgiaccio) [reply webapp/loggedin/modules/users/userInvite/userInviteView.js]: no... deleted

File: webapp/loggedin/modules/users/userList/userListView.js (hunk 14 above)
Comment (sgiaccio) [reply webapp/loggedin/modules/users/userList/userListView.js]: ok

File: server/system/aws.js (hunk 11 above)
Comment (sgiaccio) [reply server/system/aws.js]: We might want to rotate them in the future but for now setting them just once

File: server/modules/activityLog/activityLogger.js (hunk 5 above)
Comment (sgiaccio) [reply server/modules/activityLog/activityLogger.js]: ok

File: server/modules/user/manager/userManager.js [hunk 15]
```diff
@@ -1,35 +1,60 @@
...
+const AuthManager = require('../../auth/manager/authManager')
+
+const User = require('../../../../common/user/user')
+
+const ActivityLog = require('../../activityLog/activityLogger')
+
+// ==== CREATE
+
+const insertUser = async (user, surveyId, email, groupId, client = db) =>
+  client.tx(async t => {
```
Comment (minotogna) [inline server/modules/user/manager/userManager.js]: you should never forget await .

File: server/modules/user/manager/userManager.js [hunk 16]
```diff
@@ -1,35 +1,60 @@
...
+    const newUser = await UserRepository.insertUser(surveyId, email, groupId, t)
+    await ActivityLog.log(user, surveyId, ActivityLog.type.userJoined, { email, id: User.getId(newUser) }, t)
+
```
Comment (minotogna) [inline server/modules/user/manager/userManager.js]: has here the user accepted the invitation? ActivityLog.type.userJoined should be inserted when the user accept the invitation.

File: server/modules/user/manager/userManager.js [hunk 17]
```diff
@@ -1,35 +1,60 @@
...
+    await AuthManager.addUserToGroup(user, surveyId, groupId, User.getId(newUser), t)
+
```
Comment (minotogna) [inline server/modules/user/manager/userManager.js]: we use the convention that Manager uses Repositories

File: server/system/aws.js (hunk 11 above)
Comment (minotogna) [reply server/system/aws.js]: where is the config happening now? I couldn't see it..

**Pull request merged**